### ISSUE TITLE
Discussions regarding QUIC and TLS in general

### ISSUE THREAD

**Issue by Mygod** [v2ray/discussion#113]
This moves discussions from https://github.com/shadowsocks/shadowsocks-org/issues/139 starting [here](https://github.com/shadowsocks/shadowsocks-org/issues/139#issuecomment-453053217). (it was kind of off topic)

For completeness, I will recap my points here. Let me know what you think. 😄 

## Obfuscations

So after discussions from that thread, I personally think that there are three reasonable combinations to use: plain (obfuscates as QUIC traffic), encrypted (obfuscates as completely random packets, I suppose), or encrypted + other obfuscation (obfuscates as some other specific kind of traffic). I don't see how using other obfuscation only can be useful as I believe private-key encryption overhead is negligible. So to me, I think it's best if the documentation can tell people to simply avoid using obfuscation only.

## SNI and Certificate Generations

Currently auto-generated certificates will be signed for `quic.internal.v2ray.com`. This is problematic as I believe [v2ray.com is blocked in China](https://zh.greatfire.org/https/v2ray.com) and this will make it very easy to discover v2ray servers with such certificates using active probing (this technique is very cheap to deploy). A better hostname to hardcode I imagine could be `github.com` or `microsoft.com`.

But I suggest it would be best if users can configure the hostname they want (I believe this shouldn't be too hard to do with golang), and recommend users to use an actually valid hostname, e.g. `stream.mypersonal.blog` if the user happens to own the domain `mypersonal.blog`. A good default with less potential collateral damage can also be server's public IP address, for example see https://1.1.1.1.

(basically, I don't think unencrypted SNI is a big issue, the motivation for proposing encrypted SNI is mostly regarding privacy)

## Misc

I also have a smaller suggestion for doc. For example, I see `v2ray-core` uses [lucas-clemente](https://github.com/lucas-clemente/quic-go)'s implementation of QUIC. I think the doc should probably mention this as there are so many QUIC variants (IETF drafts, Google, etc.) as of now.

**Comment by VictoriaRaymond** [comment]
> I think it's best if the documentation can tell people to simply avoid using obfuscation only.

V2Ray is not opinionated. It doesn't tell users what to do and what not. Actually there are many combinations in V2Ray that are not "recommended". It is users' responsibility to [pick a solution that 
works best for them](https://steemit.com/cn/@v2ray/47wumq).

> This is problematic as I believe v2ray.com is blocked in China

As mentioned in this other thread, user can generate a TLS certificate by him/herself, with a different domain name. It is a supported feature in V2Ray. The auto-generation is a fast path for users who don't wanna do that.

> the doc should probably mention this

IEFT Quic is used in V2Ray. This is mentioned in doc.

**Comment by Mygod** [comment]
I agree that having more options is a good thing, especially under this context, but in reality most users won't be bothered and will probably just follow what others do. I believe a good default or a recommended configuration (or a few) is good for users as a starting point for their customization.

On the other hand though, IMO there are also a few options that are painfully easy to identify as designed for bypassing censorship, that it should probably be discouraged to say the least. This is the case for the 2 points I raised earlier. Obfuscation-only is a bad idea because double obfuscation equals no obfuscation. I acknowledge that the user can pick their own TLS certificates but the current auto generate functionality is misleading (I believe there will be a good amount of users that will simply resort this).

Basically, my point is that we should probably tell user to avoid option A if there exists option B that is better than A in basically every way. I believe this shouldn't contradict your core design philosophy.

> IEFT Quic is used in V2Ray. This is mentioned in doc.

The point is that IETF QUIC is still in draft state. So implementations differ depending on which draft they are implementing. It's probably best to include the full link IMO.

**Comment by VictoriaRaymond** [comment]
Your assumption is, this tool is used to fight against "the" firewall (possibly the best one). But it is not true. V2Ray can be used in other simpler scenarios, e.g, changing source IP to visit Netflix. In these cases, extra encryption is not necessary at all.

There are [docs for beginners](https://toutyrater.github.io/). We highly appreciate such docs, and periodically review them to make sure its content are up to date.

As for Quic, both spec and implementation are changing rapidly these days. It is difficult to document the current state. We will update doc once it reaches a stable state.

**Comment by Mygod** [comment]
In that case, as I mentioned before, based on my experiences private-key operations (i.e. extra encryption) are so cheap on modern CPUs that it just won't be a bottleneck for virtually everything. As for simpler scenarios (like Netflix or simply changing IP), I would prefer something much more lightweight like SOCKS proxy or WireGuard, just like I wouldn't use Microsoft Word or LaTeX to say write READMEs. I hope some people would find V2Ray useful and worth the trouble.

I think the quickest way to document the current state for QUIC is to add a link to the underlying implementation used, but that's up to you.

**Comment by VictoriaRaymond** [comment]
You are welcome to write a blogpost and explain your decision choices. Personally I believe your choices are reasonable. It is just not the interest of office doc.

**Comment by Mygod** [comment]
I think it's good enough to keep the discussions here as I believe I have already (hopefully) made all my points clear. I guess I can understand why you choose not to take these suggestions, although I still hope to have a better default for certificate generation.

Also here's how caddy [handles automatic certificate generation](https://caddyserver.com/docs/automatic-https) and [documents QUIC](https://github.com/mholt/caddy/wiki/QUIC).

**Comment by klzgrad** [comment]
I'd like to interject for a moment.

> V2Ray is not opinionated. It doesn't tell users what to do and what not. Actually there are many combinations in V2Ray that are not "recommended".

What this effectively does is put the job of protocol design in the hands of the user. The result: they don't know what is best.

This combo already starts to look bizarre. Shadowsocks rolls its own cryptos, and V2 does that too, what then becomes Shadowsocks on top of V2? Now you realize there should be a single layer for cryptos and that's why people settle on putting cryptos into TLS. Also, V2 rolls its multiplexing too while HTTP/2 already does it.

But this might where "worse is better" becomes true, as cross-pollination of crap protocol combinations makes it a pain in the ass to program traffic detectors.
